Add per thumb Range Slider semantics

guidezpl opened the pull request:
## Description

Each `RangeSlider` thumb now has its own `SemanticNode`, each controllable like an individual slider, and which respects gesture rules like `minThumbSeparation` and inability to pass the other.

The `semanticFormatterCallback` passed to `RangeSlider` now formats both values individually instead of together (e.g. '10$' and '25$' instead of '15$').

In action: https://i.imgur.com/0ZEfYyF.mp4

## Related Issues

Fixes https://github.com/flutter/flutter/issues/60715
#39510 

## Tests

A test to validate the a RangeSlider's semantics tree 

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

guidezpl committed:
Add per thumb Range Slider semantics

guidezpl committed:
Fix missing semantics if disabled

rami-a commented on packages/flutter/lib/src/material/range_slider.dart:
```diff
@@ -1513,37 +1514,93 @@ class _RenderRangeSlider extends RenderBox with RelayoutWhenSystemFontsChangeMix
+  @override
+  void describeSemanticsConfiguration(SemanticsConfiguration config) {
+    super.describeSemanticsConfiguration(config);
+
```
Nit: Remove new line

rami-a commented on packages/flutter/lib/src/material/range_slider.dart:
```diff
@@ -858,6 +858,8 @@ class _RenderRangeSlider extends RenderBox with RelayoutWhenSystemFontsChangeMix
 
   bool get isDiscrete => divisions != null && divisions > 0;
 
+  double get minThumbSeparationValue => isDiscrete ? 0 : sliderTheme.minThumbSeparation / _trackRect.width;
```
Nit: I think this can be private

rami-a commented on packages/flutter/lib/src/material/range_slider.dart:
```diff
@@ -1513,37 +1514,93 @@ class _RenderRangeSlider extends RenderBox with RelayoutWhenSystemFontsChangeMix
+      config.increasedValue = semanticFormatterCallback(_state._lerp((value + _semanticActionUnit).clamp(0.0, 1.0) as double));
+      config.decreasedValue = semanticFormatterCallback(_state._lerp((value - _semanticActionUnit).clamp(0.0, 1.0) as double));
+    } else {
+      config.value = '${(value * 100).round()}%';
```
Should percent be used here? If so, what about localization?

guidezpl replied:
This matches what `Slider` currently does. If they are to be localized, that would have to be a separate PR

clocksmith commented on packages/flutter/lib/src/material/range_slider.dart:
```diff
@@ -1513,37 +1514,93 @@ class _RenderRangeSlider extends RenderBox with RelayoutWhenSystemFontsChangeMix
+    );
+
+    // Split the semantics node area between the start and end nodes.
+    _startSemanticsNode.rect = Rect.fromPoints(node.rect.topLeft, node.rect.bottomCenter);
```
Will these be reversed for RTL?

guidezpl replied:
Good catch, I'll reverse them for RTL

guidezpl committed:
Address feedback

guidezpl committed:
Reverse nodes' rect when RTL

goderbauer commented on packages/flutter/lib/src/material/range_slider.dart:
```diff
@@ -1513,37 +1514,102 @@ class _RenderRangeSlider extends RenderBox with RelayoutWhenSystemFontsChangeMix
+  void assembleSemanticsNode(
+      SemanticsNode node,
+      SemanticsConfiguration config,
+      Iterable<SemanticsNode> children,
```
assert that children is empty since the method ignores it?

goderbauer commented on packages/flutter/test/material/range_slider_test.dart:
```diff
@@ -1742,6 +1743,86 @@ void main() {
+
+    await tester.pumpAndSettle();
+
+    expect(
```
I'd recommend writing the test with `tester.getSemantics` and `matchesSemantics` (example: https://github.com/flutter/flutter/pull/60329/files). That makes the test a little less brittle (we don't have to update the test when semantics above the Slider change.

guidezpl replied:
Cool, I'll look into that. It feels more resilient

goderbauer commented on packages/flutter/lib/src/material/range_slider.dart:
```diff
@@ -1513,37 +1514,102 @@ class _RenderRangeSlider extends RenderBox with RelayoutWhenSystemFontsChangeMix
+    );
+
+    // Split the semantics node area between the start and end nodes.
+    final Rect leftRect = Rect.fromPoints(node.rect.topLeft, node.rect.bottomCenter);
+    final Rect rightRect = Rect.fromPoints(node.rect.topCenter, node.rect.bottomRight);
```
This means, the accessibility focus rect (that green rect that Talkback draws) will always cover the first half of the slider for the first thump and the second half of the slider for the second thump - no matter the thumb position, right? Is that the expected behavior for a range slider? I would have expected the semantics nodes to be around the thumbs...

guidezpl replied:
Correct, the thinking was that splitting the area in two makes the widget more accessible due to the larger targets (the slider widget's semantic node is around the whole slider, for instance). It also helps when the two sliders are on top of one another

guidezpl committed:
Replace useSemantics in test with matchesSemantics

guidezpl committed:
Assert empty children

guidezpl committed:
Remove trailing space

goderbauer reviewed (approved):
LGTM

guidezpl committed:
Remove unused import

guidezpl committed:
Simplify increase/decrease value logic

guidezpl committed:
Update RangeSlider test to check for close thumbs

guidezpl committed:
Fix floating point bug

guidezpl commented:
Fixed and simplified the increase/decrease logic to avoid cases where the min distance between two close thumbs when using semantic actions was determined by the min thumb separation.

In order words, the thumbs now always get adjusted by the same increment (platform-determined for continuous sliders, user-specified for discrete sliders), or not at all.
